v2.5.0

arbrandes opened the pull request:
Introduces multi-provider support.

fghaas commented:
Ah no worries, I won't. :)

arbrandes commented:
Ok, this is ready for testing and code review.

fghaas commented on README.md:
```diff
@@ -380,6 +377,17 @@ configured with the following attributes:
 
+You can also use the following nested XML tags:
+
+* `provider`: (Optional) A list of OpenStack providers configured in the
```
Minor nit: this sounds like the xblock expects a _single_ nested `<provider>` tag, which is then interpreted as a list. I'd suggest to rephrase this as:

```markdown
* `provider`: Reference to an OpenStack provider configured in the platform.
  It `name` attribute must match one of the providers in the XBlock configuration,
  `capacity` specifies how many environments should be launched in that
  provider at maximum (where 0 means keep launching environments until 
  encountering a launch failure), and 
  `environment` specifies a content store path to a Heat environment file specifying
  template parameters for that provider.
  This element must be specified at least once, and can be specified multiple times
  to use one or more fallback providers.
```

The part where it says "where 0 means launch until encountering a launch failure" is of course subject to the rest of the discussion, i.e. what does setting 0 mean, as opposed to not defining the attribute at all. Is it equivalent (meaning 0 means launch until failure), or is it 0 actually 0 (meaning launch nothing here at all, something that one might use to temporarily disable a provider in a deployed course) .

arbrandes replied:
+1 for improving the wording.  As for 0 versus `None`, they're currently equivalent, in that they translate to infinite capacity.  However, you have a point:  we may want to use -1 for infinite, and 0 for disabled.  What do you think?

fghaas commented on README.md:
```diff
@@ -380,6 +377,17 @@ configured with the following attributes:
   shown as a dropdown to the user.  It must match the list of port numbers in
   both order and number.
 
+You can also use the following nested XML tags:
```
Suggest to use "elements" instead of "tags" here.

fghaas commented on README.md:
```diff
@@ -380,6 +377,17 @@ configured with the following attributes:
+  specifies a content store path to a Heat environment file specifying template
+  parameters for that provider.
+
+* `test`: (Optional) A list of scripts that will run against the lab
```
Same as above. Suggest to clarify that you can define _multiple_ `<test>` elements which then _constitute_ a list of tests, rather than a _single_ `<test>` element that _contains_ a list of tests.

fghaas commented on hastexo/hastexo.py:
```diff
@@ -126,6 +133,17 @@ def parse_xml(cls, node, runtime, keys, id_generator):
+                name = child.attrib["name"]
+                if not name:
+                    raise KeyError("name")
+                capacity = child.attrib.get("capacity", None)
```
Since this defaults to `None`, what is the defined behavior if the user sets `0`, or a negative number?

fghaas commented on hastexo/hastexo.py:
```diff
@@ -163,31 +181,25 @@ def get_block_ids(self):
-            stack_template = asset.data
+            contents = asset.data
         except (ImportError, NotFoundError):
             pass
```
Do we really want to succeed silently on failure here? Not even a log message if the course author fails to include a template, or botches the reference?

fghaas commented on hastexo/hastexo.py:
```diff
@@ -318,24 +299,63 @@ def launch_stack_task_result(self, task_id):
-                self.stack_user_name,
-                reset
-            )
+            stack_template = self.read_from_contentstore(
```
What happens if this read fails? I see no exception handler here, which is generally fine — we do want to bail if the template can't be found. But would we get a stack trace somewhere?

fghaas commented on hastexo/hastexo.py:
```diff
@@ -318,24 +299,63 @@ def launch_stack_task_result(self, task_id):
+                for provider in self.providers:
+                    p = dict(provider)
+                    env_path = p.get("environment")
+                    p["environment"] = self.read_from_contentstore(env_path)
```
Same as above: how do we handle a read failure here?

fghaas commented on hastexo/tasks.py:
```diff
@@ -76,174 +74,332 @@ def run(self,
+
+        return stack_data
+
+    def get_provider_stack_count(self, provider):
```
I very much like the idea that instead of keeping track of a "stacks per provider" semaphore count, we just do a `SELECT COUNT` every time we need to get the stack count. We'll need to see how this scales, I suppose, but my hunch is that this scales far better than any contentious updates.

arbrandes replied:
Yup, my feelings exactly.  However, I don't _think_ this will bog down our mariadb cluster any time soon.  The `hastexo_stack` table has not yet gone beyond 10k records in any of our production stacks, and unless we're talking tens of millions of stacks, we can probably handle our worst cases with some further InnoDB tuning.

arbrandes commented:
I've addressed the comments in the latest fixup [commit](https://github.com/hastexo/hastexo-xblock/pull/79/commits/66eb7bad2c594ce064871bce4bbeb936d66f9893), which I'll squash later.

fghaas commented on README.md:
```diff
@@ -377,17 +377,22 @@ configured with the following attributes:
+  specified at least once, and can be specified multiple times to use one or
+  more fallback providers.
+
+* `test`: The contents of this element will be run verbatum as a script in the
```
It's "verbatim". :)

fghaas commented on README.md:
```diff
@@ -377,17 +377,22 @@ configured with the following attributes:
+
+* `test`: The contents of this element will be run verbatum as a script in the
+  user's lab environment, when the they click the "Check Progress" button.  As
+  such, it should define an interpreter via the "shabang" convention.  This
```
Commonly "shebang".

fghaas commented on hastexo/tasks.py:
```diff
@@ -175,12 +175,22 @@ def try_all_providers(self):
+                try:
+                    capacity = int(capacity)
+                except (TypeError, ValueError):
+                    logger.info("Provider [%s] has invalid capacity." %
```
`.info()` seems weak here. This should be a warning or an error, I think.

arbrandes commented:
All good points. :)  Addressed [here](https://github.com/hastexo/hastexo-xblock/pull/79/commits/584d6663e26c3981f4b6f191aec21a5a2a1ae0b0)

arbrandes commented:
Any objections to merging this in?

fghaas commented:
None, go ahead!
